compaction/compaction_strategy.cc

```diff
@@ -168,26 +170,36 @@ double size_tiered_backlog_tracker::backlog(const compaction_backlog_tracker::on
     auto effective_backlog_bytes = total_backlog_bytes - compacted.total_bytes;
 
     // Sum of (Si - Ci) * log (Si) for all SSTables contributing backlog
-    auto sstables_contribution = _sstables_backlog_contribution - compacted.contribution;
+    auto sstables_contribution = _contrib.value - compacted.contribution;
     // This is subtracting ((Si - Ci) * log (Si)) from ((Si - Ci) * log(T)), yielding the final backlog
-    auto b = (effective_backlog_bytes * log4(_total_bytes)) - sstables_contribution;
+    auto b = (effective_backlog_bytes * utils::log_base<4>(_total_bytes)) - sstables_contribution;
     return b > 0 ? b : 0;
 }
 
-void size_tiered_backlog_tracker::replace_sstables(std::vector<sstables::shared_sstable> old_ssts, std::vector<sstables::shared_sstable> new_ssts) {
+// FIXME: Should provide strong exception safety guarantees
+void size_tiered_backlog_tracker::replace_sstables(const std::vector<sstables::shared_sstable>& old_ssts, const std::vector<sstables::shared_sstable>& new_ssts) {
+    int64_t delta_bytes = 0;
+    auto all = _all;
+    all.reserve(_all.size() - old_ssts.size() + new_ssts.size());
+
```

Comment:
I really dislike that we're managing the set of sstables in the tracker.

Backlog tracker will eventually land on compaction manager. Table doesn't have to be aware of it. 

I am planning to wire compaction strategy logic with the backlog tracker, so we can reuse the get_buckets() logic which is N log N on number of sstables. For inserting N sstables, you fall into quadratic behavior. The impact of this is reduced as N is small for STCS, and that we're doing a single contribution recalculation on every replacement.

If the tracker is no longer maintaining the sstable set, exception handling becomes simpler.

A failure to calculate backlog can be corrected in the next attempt. The problem was living with a partial sstable set, which doesn't reflect the one in replica::table. Having two sources of truth for the sstable set is asking for trouble.

I suggest we drop this series in favor of a revamp.

Reply:
Sometimes best is the enemy of good.
Improving exception safety with what we have doesn't prevent us from improving it in the future.

Also, achieving exception safety with the current design is feasible to bavkport, while I presume that the revamped code wouldn't be.

Stratigically, I'd like to proceed in smaller steps, so to establish a true, strongly exception safe framework, even if it's not optimal with respect to memory allocations and complexity, and then continue to improve it within the safe boundaries we've set.

Reply:
I agree that it's not good for the backlog tracker to maintain its own sstable set. Soon we'll start to worry about whether the sets match.

Perhaps a way to enforce consistency is with an undo record. replace_sstables() returns an undo record that undoes the work, unless cancelled.

Reply:
The problem with the undo record is that if we want it to be efficient and undo the individual erases and inserts, it might throw in the undo path which should be noexcept.  For a wholesale undo that copies the input before modifying it, its efficiency would be the same as making the changes on temporary copies and making sure that applying them will not throw.

Reply:
std::map has extract/merge members which can temporarily remove nodes with the ability to merge them back again. But it's a large project to apply the concept to our custom data structures, and maybe impossible as long as we use the interval map.